File: app/src/main/java/com/mask/customcomponents/view/DividerGridItemDecoration.java
package com.mask.customcomponents.view;

import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.OrientationHelper;
import androidx.recyclerview.widget.RecyclerView;

/**
 * RecyclerView网格间距
 * <p>
 * Created by lishilin on 2016/11/15
 */
public class DividerGridItemDecoration extends RecyclerView.ItemDecoration {

    private static final Drawable dividerDefault = new ColorDrawable(0x00000000);
    private static final int dividerWidthDefault = 0;
    private static final int dividerHeightDefault = 0;

    private Drawable divider;
    private int dividerWidth;
    private int dividerHeight;

    private boolean includeEdge;// 是否包含边缘(垂直布局 - 是否有上下间距，水平布局 - 是否有左右间距)

    public static DividerGridItemDecoration getInstance() {
        return getInstance(dividerWidthDefault, dividerHeightDefault);
    }

    public static DividerGridItemDecoration getInstance(int width, int height) {
        return getInstance(width, height, dividerDefault);
    }

    public static DividerGridItemDecoration getInstance(int width, int height, @ColorInt int color) {
        return getInstance(width, height, new ColorDrawable(color));
    }

    public static DividerGridItemDecoration getInstance(int width, int height, Drawable divider) {
        final DividerGridItemDecoration decoration = new DividerGridItemDecoration();
        decoration.setDivider(divider);
        decoration.setDividerHeight(width);
        decoration.setDividerWidth(height);
        return decoration;
    }

    private DividerGridItemDecoration() {
        super();
    }

    @Override
    public void onDraw(@NonNull Canvas c, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        if (divider == null) {
            return;
        }

        final int orientation = getOrientation(parent);

        if (orientation == OrientationHelper.VERTICAL) {
            drawVerticalDividerHorizontal(c, parent);
            drawVerticalDividerVertical(c, parent);
        } else if (orientation == OrientationHelper.HORIZONTAL) {
            drawHorizontalDividerHorizontal(c, parent);
            drawHorizontalDividerVertical(c, parent);
        }
    }

    @Override
    public void getItemOffsets(@NonNull Rect outRect, @NonNull View view, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        final int orientation = getOrientation(parent);

        if (orientation == OrientationHelper.VERTICAL) {
            if (isFirstLine(parent, view) && includeEdge) {// 绘制上间距(当 是第一排 且 包含边缘)
                outRect.top = dividerHeight;
            }

            if (!isLastLine(parent, view) || includeEdge) {// 绘制下间距(当 不是最后一排 或 包含边缘)
                outRect.bottom = dividerHeight;
            }

            outRect.left = getVerticalLeftSpace(parent, view);// 绘制左间距
            outRect.right = getVerticalRightSpace(parent, view);// 绘制右间距
        } else if (orientation == OrientationHelper.HORIZONTAL) {
            if (isFirstLine(parent, view) && includeEdge) {// 绘制左间距(当 是第一排 且 包含边缘)
                outRect.left = dividerWidth;
            }

            if (!isLastLine(parent, view) || includeEdge) {// 绘制右间距(当 不是最后一排 或 包含边缘)
                outRect.right = dividerWidth;
            }

            outRect.top = getHorizontalTopSpace(parent, view);// 绘制上间距
            outRect.bottom = getHorizontalBottomSpace(parent, view);// 绘制下间距
        }
    }

    /**
     * 绘制垂直滚动横向线条
     *
     * @param c      Canvas
     * @param parent RecyclerView
     */
    private void drawVerticalDividerHorizontal(Canvas c, RecyclerView parent) {
        final int childCount = parent.getChildCount();

        for (int childViewIndex = 0; childViewIndex < childCount; childViewIndex++) {
            final View child = parent.getChildAt(childViewIndex);

            if (isFirstLine(parent, child) && includeEdge) {// 绘制上间距(当 是第一排 且 包含边缘)
                int left = child.getLeft();
                int top = child.getTop() - dividerHeight;
                int right = left + child.getWidth() + dividerWidth;
                int bottom = top + dividerHeight;

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }

            if (!isLastLine(parent, child) || includeEdge) {// 绘制下间距(当 不是最后一排 或 包含边缘)
                int left = child.getLeft();
                int top = child.getBottom();
                int right = left + child.getWidth() + dividerWidth;
                int bottom = top + dividerHeight;

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }
        }
    }

    /**
     * 绘制垂直滚动纵向线条
     *
     * @param c      Canvas
     * @param parent RecyclerView
     */
    private void drawVerticalDividerVertical(Canvas c, RecyclerView parent) {
        final int childCount = parent.getChildCount();

        for (int childViewIndex = 0; childViewIndex < childCount; childViewIndex++) {
            final View child = parent.getChildAt(childViewIndex);

            final int dividerWidthLeft = getVerticalLeftSpace(parent, child);
            final int dividerWidthRight = getVerticalRightSpace(parent, child);

            if (dividerWidthLeft > 0) {// 绘制左间距
                int left = child.getLeft() - dividerWidthLeft;
                int top = child.getTop();
                int right = left + dividerWidthLeft;
                int bottom = top + child.getHeight();

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }

            if (dividerWidthRight > 0) {// 绘制右间距
                int left = child.getRight();
                int top = child.getTop();
                int right;
                // 如果是最后一个且未占满一排(后面有空白区域)
                if (childViewIndex == childCount - 1) {
                    right = left + dividerWidth;
                } else {
                    right = left + dividerWidthRight;
                }
                int bottom = top + child.getHeight();

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }
        }
    }

    /**
     * 绘制水平滚动横向线条
     *
     * @param c      Canvas
     * @param parent RecyclerView
     */
    private void drawHorizontalDividerHorizontal(Canvas c, RecyclerView parent) {
        final int childCount = parent.getChildCount();

        for (int childViewIndex = 0; childViewIndex < childCount; childViewIndex++) {
            final View child = parent.getChildAt(childViewIndex);

            final int dividerWidthTop = getVerticalLeftSpace(parent, child);
            final int dividerWidthBottom = getVerticalRightSpace(parent, child);

            if (dividerWidthTop > 0) {// 绘制上间距
                int left = child.getLeft();
                int top = child.getTop() - dividerWidthTop;
                int right = left + child.getWidth();
                int bottom = top + dividerWidthTop;

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }

            if (dividerWidthBottom > 0) {// 绘制下间距
                int left = child.getLeft();
                int top = child.getBottom();
                int right = left + child.getWidth();
                // 如果是最后一个且未占满一排(后面有空白区域)
                int bottom;
                if (childViewIndex == childCount - 1) {
                    bottom = top + dividerHeight;
                } else {
                    bottom = top + dividerWidthBottom;
                }

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }
        }
    }

    /**
     * 绘制水平滚动纵向线条
     *
     * @param c      Canvas
     * @param parent RecyclerView
     */
    private void drawHorizontalDividerVertical(Canvas c, RecyclerView parent) {
        final int childCount = parent.getChildCount();

        for (int childViewIndex = 0; childViewIndex < childCount; childViewIndex++) {
            final View child = parent.getChildAt(childViewIndex);

            if (isFirstLine(parent, child) && includeEdge) {// 绘制左间距(当 是第一排 且 包含边缘)
                int left = child.getLeft() - dividerWidth;
                int top = child.getTop();
                int right = left + dividerWidth;
                int bottom = top + child.getHeight() + dividerHeight;

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }

            if (!isLastLine(parent, child) || includeEdge) {// 绘制右间距(当 不是最后一排 或 包含边缘)
                int left = child.getRight();
                int top = child.getTop();
                int right = left + dividerWidth;
                int bottom = top + child.getHeight() + dividerHeight;

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }
        }
    }

    /**
     * 获取列数
     *
     * @param parent RecyclerView
     * @return 列数
     */
    private int getSpanCount(RecyclerView parent) {
        int spanCount = -1;
        RecyclerView.LayoutManager layoutManager = parent.getLayoutManager();
        if (layoutManager instanceof GridLayoutManager) {
            spanCount = ((GridLayoutManager) layoutManager).getSpanCount();
        }
        return spanCount;
    }

    /**
     * 获取当前View列下标
     *
     * @param parent RecyclerView
     * @param view   view
     * @return 列下标
     */
    private int getSpanIndex(RecyclerView parent, View view) {
        int spanIndex = -1;
        ViewGroup.LayoutParams params = view.getLayoutParams();
        if (params instanceof GridLayoutManager.LayoutParams) {
            final int spanCount = getSpanCount(parent);
            final int position = getChildPosition(parent, view);// item position
            spanIndex = position % spanCount;// item spanIndex
        }
        return spanIndex;
    }

    /**
     * 获取方向
     *
     * @param parent RecyclerView
     * @return 方向
     */
    private int getOrientation(RecyclerView parent) {
        int orientation = -1;
        RecyclerView.LayoutManager layoutManager = parent.getLayoutManager();
        if (layoutManager instanceof GridLayoutManager) {
            orientation = ((GridLayoutManager) layoutManager).getOrientation();
        }
        return orientation;
    }

    /**
     * 获取位置position
     *
     * @param parent RecyclerView
     * @param view   child
     * @return position
     */
    private int getChildPosition(RecyclerView parent, View view) {
        return parent.getChildAdapterPosition(view);
    }

    /**
     * 获取垂直滑动item左间距
     *
     * @param parent RecyclerView
     * @param view   item
     * @return 垂直滑动item左间距
     */
    private int getVerticalLeftSpace(RecyclerView parent, View view) {
        final int spanCount = getSpanCount(parent);
        final int spanIndex = getSpanIndex(parent, view);// item spanIndex

        return spanIndex * dividerWidth / spanCount;
    }

    /**
     * 获取垂直滑动item右间距
     *
     * @param parent RecyclerView
     * @param view   item
     * @return 垂直滑动item右间距
     */
    private int getVerticalRightSpace(RecyclerView parent, View view) {
        final int spanCount = getSpanCount(parent);
        final int spanIndex = getSpanIndex(parent, view);// item spanIndex

        return dividerWidth - (spanIndex + 1) * dividerWidth / spanCount;
    }

    /**
     * 获取水平滑动item上间距
     *
     * @param parent RecyclerView
     * @param view   item
     * @return 水平滑动item上间距
     */
    private int getHorizontalTopSpace(RecyclerView parent, View view) {
        final int spanCount = getSpanCount(parent);
        final int spanIndex = getSpanIndex(parent, view);// item spanIndex

        return spanIndex * dividerHeight / spanCount;
    }

    /**
     * 获取水平滑动item下间距
     *
     * @param parent RecyclerView
     * @param view   item
     * @return 水平滑动item下间距
     */
    private int getHorizontalBottomSpace(RecyclerView parent, View view) {
        final int spanCount = getSpanCount(parent);
        final int spanIndex = getSpanIndex(parent, view);// item spanIndex

        return dividerHeight - (spanIndex + 1) * dividerHeight / spanCount;
    }

    /**
     * 是否是第一排(垂直布局 - 第一横排，水平布局 - 第一竖排)
     *
     * @param parent RecyclerView
     * @param view   当前view
     * @return 是否是第一排
     */
    private boolean isFirstLine(RecyclerView parent, View view) {
        final int position = getChildPosition(parent, view);// item position
        final int spanCount = getSpanCount(parent);

        return position < spanCount;
    }

    /**
     * 是否是最后一排(垂直布局 - 最后一横排，水平布局 - 最后一竖排)
     *
     * @param parent RecyclerView
     * @param view   当前view
     * @return 是否是最后一排
     */
    private boolean isLastLine(RecyclerView parent, View view) {
        RecyclerView.Adapter adapter = parent.getAdapter();
        final int position = getChildPosition(parent, view);// item position
        final int spanCount = getSpanCount(parent);
        int childCount = adapter == null ? 0 : adapter.getItemCount();// item总数
        int lastLineCount = childCount % spanCount;// 最后一排item数量

        if (childCount <= spanCount) {
            return true;
        }

        if (lastLineCount == 0) {
            lastLineCount = spanCount;
        }

        childCount = childCount - lastLineCount;
        return position >= childCount;
    }

    public boolean isIncludeEdge() {
        return includeEdge;
    }

    public void setIncludeEdge(boolean includeEdge) {
        this.includeEdge = includeEdge;
    }

    /**
     * Sets the drawable to be used as the divider.
     *
     * @param divider divider
     */
    public void setDivider(Drawable divider) {
        if (divider != null) {
            int intrinsicWidth = divider.getIntrinsicWidth();
            int intrinsicHeight = divider.getIntrinsicHeight();
            if (intrinsicWidth > 0) {
                dividerWidth = intrinsicWidth;
            }
            if (intrinsicHeight > 0) {
                dividerHeight = intrinsicHeight;
            }
        } else {
            dividerWidth = 0;
            dividerHeight = 0;
        }
        this.divider = divider;
    }

    /**
     * Gets the drawable currently used as the divider.
     */
    public Drawable getDivider() {
        return divider;
    }

    /**
     * Sets the divider width, in pixels.
     */
    public void setDividerWidth(int dividerWidth) {
        this.dividerWidth = dividerWidth;
    }

    /**
     * Gets the divider width, in pixels.
     */
    public int getDividerWidth() {
        return dividerWidth;
    }

    /**
     * Sets the divider height, in pixels.
     */
    public void setDividerHeight(int dividerHeight) {
        this.dividerHeight = dividerHeight;
    }

    /**
     * Gets the divider height, in pixels.
     */
    public int getDividerHeight() {
        return dividerHeight;
    }

}

File: app/src/main/java/com/mask/customcomponents/decoration/DividerGridItemDecoration.kt
package com.mask.customcomponents.decoration

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.view.View
import androidx.annotation.ColorInt
import androidx.core.graphics.drawable.toDrawable
import androidx.core.view.forEach
import androidx.core.view.size
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.OrientationHelper
import androidx.recyclerview.widget.RecyclerView

/**
 * RecyclerView GridLayoutManager 网格分隔线
 *
 * Created by lishilin on 2016/11/15
 */
class DividerGridItemDecoration private constructor(
    divider: Drawable?,
    dividerWidth: Int,
    dividerHeight: Int
) : RecyclerView.ItemDecoration() {

    private val mDivider = divider ?: mDividerDefault
    private val mDividerWidth = dividerWidth
    private val mDividerHeight = dividerHeight

    var isIncludeEdge = false // 是否包含边缘(垂直布局：是否有上下分隔线； 水平布局：是否有左右分隔线；)

    companion object {

        private val mDividerDefault: Drawable
            get() {
                return Color.TRANSPARENT.toDrawable()
            }

        fun newInstance(width: Int, height: Int): DividerGridItemDecoration {
            return newInstance(width, height, mDividerDefault)
        }

        fun newInstance(width: Int, height: Int, @ColorInt color: Int): DividerGridItemDecoration {
            return newInstance(width, height, color.toDrawable())
        }

        fun newInstance(width: Int, height: Int, divider: Drawable?): DividerGridItemDecoration {
            return DividerGridItemDecoration(divider, width, height)
        }
    }

    override fun getItemOffsets(
        outRect: Rect,
        view: View,
        parent: RecyclerView,
        state: RecyclerView.State
    ) {
        when (getOrientation(parent)) {
            OrientationHelper.VERTICAL -> {
                getItemOffsetsByVertical(outRect, view, parent)
            }

            OrientationHelper.HORIZONTAL -> {
                getItemOffsetsByHorizontal(outRect, view, parent)
            }
        }
    }

    override fun onDraw(canvas: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        if (mDivider.alpha == 0) {
            return
        }
        when (getOrientation(parent)) {
            OrientationHelper.VERTICAL -> {
                drawHorizontalDividerByVertical(canvas, parent)
                drawVerticalDividerByVertical(canvas, parent)
            }

            OrientationHelper.HORIZONTAL -> {
                drawVerticalDividerByHorizontal(canvas, parent)
                drawHorizontalDividerByHorizontal(canvas, parent)
            }
        }
    }

    /**
     * 获取 偏移量(垂直布局)
     */
    private fun getItemOffsetsByVertical(outRect: Rect, view: View, parent: RecyclerView) {
        // 分隔线偏移量 上
        if (isRequiredDrawDividerLeftOrTop(parent, view)) {
            outRect.top = mDividerHeight
        }

        // 分隔线偏移量 下
        if (isRequiredDrawDividerRightOrBottom(parent, view)) {
            outRect.bottom = mDividerHeight
        }

        // 分隔线偏移量 左
        outRect.left = getDividerOffsetLeftByVertical(parent, view)
        // 分隔线偏移量 右
        outRect.right = getDividerOffsetRightByVertical(parent, view)
    }

    /**
     * 获取 偏移量(水平布局)
     */
    private fun getItemOffsetsByHorizontal(outRect: Rect, view: View, parent: RecyclerView) {
        // 分隔线偏移量 左
        if (isRequiredDrawDividerLeftOrTop(parent, view)) {
            outRect.left = mDividerWidth
        }

        // 分隔线偏移量 右
        if (isRequiredDrawDividerRightOrBottom(parent, view)) {
            outRect.right = mDividerWidth
        }

        // 分隔线偏移量 上
        outRect.top = getDividerOffsetTopByHorizontal(parent, view)
        // 分隔线偏移量 下
        outRect.bottom = getDividerOffsetBottomByHorizontal(parent, view)
    }

    /**
     * 获取 分隔线偏移量 左(垂直布局)
     */
    private fun getDividerOffsetLeftByVertical(parent: RecyclerView, view: View): Int {
        val spanCount = getSpanCount(parent)
        val spanIndex = getViewSpanIndex(parent, view)
        return spanIndex * mDividerWidth / spanCount
    }

    /**
     * 获取 分隔线偏移量 右(垂直布局)
     */
    private fun getDividerOffsetRightByVertical(parent: RecyclerView, view: View): Int {
        val spanCount = getSpanCount(parent)
        val spanIndex = getViewSpanIndex(parent, view)
        return mDividerWidth - (spanIndex + 1) * mDividerWidth / spanCount
    }

    /**
     * 获取 分隔线偏移量 上(水平布局)
     */
    private fun getDividerOffsetTopByHorizontal(parent: RecyclerView, view: View): Int {
        val spanCount = getSpanCount(parent)
        val spanIndex = getViewSpanIndex(parent, view)
        return spanIndex * mDividerHeight / spanCount
    }

    /**
     * 获取 分隔线偏移量 下(水平布局)
     */
    private fun getDividerOffsetBottomByHorizontal(parent: RecyclerView, view: View): Int {
        val spanCount = getSpanCount(parent)
        val spanIndex = getViewSpanIndex(parent, view)
        return mDividerHeight - (spanIndex + 1) * mDividerHeight / spanCount
    }

    /**
     * 绘制 水平分隔线(垂直布局)
     */
    private fun drawHorizontalDividerByVertical(canvas: Canvas, parent: RecyclerView) {
        // 按行遍历绘制，直接绘制一整行
        val left = 0
        val right = parent.width
        val spanCount = getSpanCount(parent)
        for (index in 0 until parent.size step spanCount) {
            val view = parent.getChildAt(index) ?: continue

            // 绘制上分隔线
            if (isRequiredDrawDividerLeftOrTop(parent, view)) {
                val top = view.top - mDividerHeight
                val bottom = top + mDividerHeight

                mDivider.setBounds(left, top, right, bottom)
                mDivider.draw(canvas)
            }

            // 绘制下分隔线
            if (isRequiredDrawDividerRightOrBottom(parent, view)) {
                val top = view.bottom
                val bottom = top + mDividerHeight

                mDivider.setBounds(left, top, right, bottom)
                mDivider.draw(canvas)
            }
        }
    }

    /**
     * 绘制 垂直分隔线(垂直布局)
     */
    private fun drawVerticalDividerByVertical(canvas: Canvas, parent: RecyclerView) {
        // 遍历绘制，最后一列不绘制
        parent.forEach { view ->
            // 绘制右分隔线
            if (!isViewLastSpan(parent, view)) {
                val left = view.right
                val right = left + mDividerWidth
                val top = view.top
                val bottom = view.bottom

                mDivider.setBounds(left, top, right, bottom)
                mDivider.draw(canvas)
            }
        }
    }

    /**
     * 绘制 垂直分隔线(水平布局)
     */
    private fun drawVerticalDividerByHorizontal(canvas: Canvas, parent: RecyclerView) {
        // 按列遍历绘制，直接绘制一整列
        val top = 0
        val bottom = parent.height
        val spanCount = getSpanCount(parent)
        for (index in 0 until parent.size step spanCount) {
            val view = parent.getChildAt(index) ?: continue

            // 绘制左分隔线
            if (isRequiredDrawDividerLeftOrTop(parent, view)) {
                val left = view.left - mDividerWidth
                val right = left + mDividerWidth

                mDivider.setBounds(left, top, right, bottom)
                mDivider.draw(canvas)
            }

            // 绘制右分隔线
            if (isRequiredDrawDividerRightOrBottom(parent, view)) {
                val left = view.right
                val right = left + mDividerWidth

                mDivider.setBounds(left, top, right, bottom)
                mDivider.draw(canvas)
            }
        }
    }

    /**
     * 绘制 水平分隔线(水平布局)
     */
    private fun drawHorizontalDividerByHorizontal(canvas: Canvas, parent: RecyclerView) {
        // 遍历绘制，最后一行不绘制
        parent.forEach { view ->
            // 绘制下分隔线
            if (!isViewLastSpan(parent, view)) {
                val left = view.left
                val right = view.right
                val top = view.bottom
                val bottom = top + mDividerHeight

                mDivider.setBounds(left, top, right, bottom)
                mDivider.draw(canvas)
            }
        }
    }

    /**
     * 获取 方向
     */
    private fun getOrientation(parent: RecyclerView): Int {
        return when (val layoutManager = parent.layoutManager) {
            is GridLayoutManager -> {
                layoutManager.orientation
            }

            else -> {
                -1
            }
        }
    }

    /**
     * 获取 列数(垂直布局)/行数(水平布局)
     */
    private fun getSpanCount(parent: RecyclerView): Int {
        return when (val layoutManager = parent.layoutManager) {
            is GridLayoutManager -> {
                layoutManager.spanCount
            }

            else -> {
                -1
            }
        }
    }

    /**
     * 获取 位置
     */
    private fun getViewPosition(parent: RecyclerView, view: View): Int {
        return parent.getChildLayoutPosition(view)
    }

    /**
     * 是否是 第一个
     */
    private fun isViewFirst(parent: RecyclerView, view: View): Boolean {
        val position = getViewPosition(parent, view)
        return position == 0
    }

    /**
     * 是否是 最后一个
     */
    private fun isViewLast(parent: RecyclerView, view: View): Boolean {
        val position = getViewPosition(parent, view)
        val childCount = parent.adapter?.itemCount ?: 0 // item总数
        return position == childCount - 1
    }

    /**
     * 获取 在当前列/行的下标
     */
    private fun getViewSpanIndex(parent: RecyclerView, view: View): Int {
        val spanCount = getSpanCount(parent)
        val position = getViewPosition(parent, view)
        return position % spanCount
    }

    /**
     * 是否是 第一个 在当前列/行
     */
    private fun isViewFirstSpan(parent: RecyclerView, view: View): Boolean {
        val spanIndex = getViewSpanIndex(parent, view)
        return spanIndex == 0
    }

    /**
     * 是否是 最后一个 在当前列/行
     */
    private fun isViewLastSpan(parent: RecyclerView, view: View): Boolean {
        val spanCount = getSpanCount(parent)
        val spanIndex = getViewSpanIndex(parent, view)
        val isViewLast = isViewLast(parent, view)
        return (spanIndex == spanCount - 1) || isViewLast
    }

    /**
     * 是否是 第一排(垂直布局：第一横排； 水平布局：第一竖排；)
     */
    private fun isViewFirstLine(parent: RecyclerView, view: View): Boolean {
        val spanCount = getSpanCount(parent)
        val position = getViewPosition(parent, view)
        return position < spanCount
    }

    /**
     * 是否是 最后一排(垂直布局：最后一横排； 水平布局：最后一竖排；)
     */
    private fun isViewLastLine(parent: RecyclerView, view: View): Boolean {
        val spanCount = getSpanCount(parent)
        val position = getViewPosition(parent, view)
        val childCount = parent.adapter?.itemCount ?: 0 // item总数
        val totalLine = (childCount + spanCount - 1) / spanCount // 总排数
        val currentLine = position / spanCount // 当前排数
        return currentLine == totalLine - 1
    }

    /**
     * 是否需要 绘制分隔线(左/上)
     *
     * 是第一排 且 包含边缘
     */
    private fun isRequiredDrawDividerLeftOrTop(parent: RecyclerView, view: View): Boolean {
        val isIncludeEdge = isIncludeEdge
        val isViewFirstLine = isViewFirstLine(parent, view)
        return isIncludeEdge && isViewFirstLine
    }

    /**
     * 是否需要 绘制分隔线(右/下)
     *
     * 不是最后一排 或 包含边缘
     */
    private fun isRequiredDrawDividerRightOrBottom(parent: RecyclerView, view: View): Boolean {
        val isIncludeEdge = isIncludeEdge
        val isViewLastLine = isViewLastLine(parent, view)
        return isIncludeEdge || !isViewLastLine
    }
}